traitlets/config/configurable.py

```diff
@@ -203,8 +210,13 @@ def _config_changed(self, change):
         section_names = self.section_names()
         self._load_config(change.new, traits=traits, section_names=section_names)
 
-    def update_config(self, config):
-        """Update config and load the new values"""
+    def update_config(self, config, skip_env=False):
```

Comment:
It is uttermost important to keep this PR backward-compatible, so I did a small study.

I [searched through all *jupyter/ipython* sources for `update_config()` invocations](https://github.com/search?utf8=%E2%9C%93&q=update_config++repo%3Aipython%2Fipython+repo%3Ajupyter%2Fdashboards+repo%3Ajupyter%2Fdocker-stacks+repo%3Ajupyter%2Fjupyter+repo%3Ajupyter%2Fjupyter_client+repo%3Ajupyter%2Fjupyter_console+repo%3Ajupyter%2Fjupyter_core+repo%3Ajupyter%2Fjupyter_markdown+repo%3Ajupyter%2Fjupyterhub+repo%3Ajupyter%2Fjupyter-kernels+repo%3Ajupyter%2Fjupyter-notebook+repo%3Ajupyter%2Fjvm-repr+repo%3Ajupyter%2Fnbconvert+repo%3Ajupyter%2Fnbdime+repo%3Ajupyter%2Fnbformat+repo%3Ajupyter%2Fnbgrader+repo%3Ajupyter%2Fnbviewer+repo%3Ajupyter%2Fnotebook+repo%3Ajupyter%2Fqt-console+extension%3Apy&type=Code&ref=advsearch&l=&l=)
and I singled out the following x4 (+ x5) `config_update()` usages:
1. [ipython/ipython@IPython/core/application.py#L462: self.update_config(<cmd-line-config>)](/ipython/ipython/blob/2118e16d74a1d4a94d73f4e0f74fbb5d3136297b/IPython/core/application.py#L462) 
   in overridden `Application.initialize()`.
2. [jupyter/jupyter_core@jupyter_core/application.py#L246: self.update_config(<cmd-line-config>)](/jupyter/jupyter_core/blob/59e09ff1c6f94d1ebbe59265af665b46a85cdbe1/jupyter_core/application.py#L246) 
   same as above.
3. [jupyter/notebook@notebook/notebookapp.py#L1091: self.update_config(<file_to_run or notebook_dir>)](/jupyter/notebook/blob/e4524a752bdcbd7083c0ba3039eec033b49e0849/notebook/notebookapp.py#L1091) 
   in overridden `Application.parse_command_line()`.
4. [ipython/ipython@IPython/core/magics/config.py#L156: configurable.update_config(<notebook-locals>)](/ipython/ipython/blob/4950ed10259c7b210cb9620c1881bb39a51ae99f/IPython/core/magics/config.py#L156) 
   in `%config` line-magic (is the "user-provided high priority calls" you talked about @minrk?)
5. x5 more usages in test-cases or other projects (e.g. nbgrader).

Now, for all of them, simply upgrading to this branch of *traitlets* would cause no harm, because:
- no `tag(envvar=...)` would exist in these downstream projects;
- any pre-existing *trait-default from `os.environ`* patterns would continue to work before, and configs would overwrite envvars.

Assuming that you Jupyter-devs want to *really* take advantage of this PR so you start adding tags to various traits.
- Current defaults of respecting envvars would benefit x5 places of (5).
- Of course (4) must NOT respect the environment, but apply strictly all user-defined locals from notebook-cell - so intervention needed.

For the rest, the situation requires intervention **for both cases.**
Specifically:

- for cases (1) & (2), the overridden `Application.initialize()` method, seems "strange",
  since it replicates the logic already present in traitlet's `Application.parse_command_line()`,
  which it already invokes, first thing at the top.
  So maybe this method needs attention anyway, not to duplicate stuff.
  But even if defaults were NOT to respect env, they would require modifications.
 
- case (3) would suffer from current defaults; but in this case i'm wondering why the call to `update_config()` is needed?
  Why not simply assign the `filename` to both traits, like this:
  
      if os.path.isdir(f):
          self.notebook_dir = f
      elif os.path.isfile(f):
          self.file_to_run = f

To conclude the situation of the x9 `update_config()` sites with current *env-wins* defaults:
- x5 are fine;
- x2 need modifications whatever defaults are choosen - and most probably need other changes as well (i.e. for not duplicating stuff or even stop overriding methods);
- x1 suffers, but can retrofit not to depend on `config_update()` at all;
- x1 suffers (the "user-provided high priority calls"),

Anyhow, choosing current defaults is 4 vs 7 changes on downstream code.
I believe backward compatibility of this PR is not a problem.